contracts/src/v0.8/shared/call/CallWithExactGas.sol

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/// @dev Handles the edge case where we want to pass a specific amount of gas,
+/// @dev but EIP-150 sends all but 1/64 of the remaining gas instead so the user gets
+/// @dev less gas than they paid for. The other 2 parts of EIP-150 do not apply since
+/// @dev a) we hard code value=0 and b) we ensure code already exists.
+/// @dev If we revert instead, then that will never happen.
+/// @dev Separately we capture the return data up to a maximum size to avoid return bombs,
+/// @dev borrowed from https://github.com/nomad-xyz/ExcessivelySafeCall/blob/main/src/ExcessivelySafeCall.sol.
+library CallWithExactGas {
+  error NoContract();
+  error NoGasForCallExactCheck();
+  error NotEnoughGasForCall();
+
+  function _callWithExactGasSafeReturnData(
+    bytes memory payload,
+    address target,
+    uint256 gasLimit,
+    uint16 maxReturnBytes,
+    uint16 gasForCallExactCheck
+  ) internal returns (bool success, bytes memory retData) {
+    // allocate retData memory ahead of time
+    retData = new bytes(maxReturnBytes);
+
+    _preCallChecks(target, gasLimit, gasForCallExactCheck);
+
+    // solhint-disable-next-line no-inline-assembly
+    assembly {
+      // call and return whether we succeeded. ignore return data
+      // call(gas,addr,value,argsOffset,argsLength,retOffset,retLength)
+      success := call(gasLimit, target, 0, add(payload, 0x20), mload(payload), 0, 0)
+
+      // limit our copy to maxReturnBytes bytes
+      let toCopy := returndatasize()
+      if gt(toCopy, maxReturnBytes) {
+        toCopy := maxReturnBytes
+      }
+      // Store the length of the copied bytes
+      mstore(retData, toCopy)
+      // copy the bytes from retData[0:_toCopy]
+      returndatacopy(add(retData, 0x20), 0, toCopy)
+    }
+    return (success, retData);
+  }
+
+  function _callWithExactGas(
+    bytes memory payload,
+    address target,
+    uint256 gasLimit,
+    uint16 gasForCallExactCheck
+  ) internal returns (bool success) {
+    _preCallChecks(target, gasLimit, gasForCallExactCheck);
```

Comment:
I understand that you want to keep it DRY, but I don't feel great about returning to solidity after entering the check logic. 
For the original `_callWithExactGas`, the check logic was intentionally in the same assembly same block as the call to ensure that I had tight control over the generated code and its gas behavior.

Reply:
Yeah I didn't want to duplicate all the logic but I'm fine duplicating it again